crates/ruma-server-util/src/authorization.rs

```diff
@@ -37,204 +43,164 @@ impl XMatrix {
-        let mut buffer = Vec::new();
-        loop {
-            match t.next()? {
-                // quoted pair
-                b'\\' => {
-                    let escaped = t.next().filter(|c| {
-                        if is_quoted_pair(**c) {
-                            true
-                        } else {
-                            debug!(
-                                "Encountered an illegal character {} at location {}",
-                                **c as char,
-                                t.location().offset()
-                            );
-                            false
-                        }
-                    })?;
-                    buffer.push(*escaped);
-                }
-                // end of quote
-                b'"' => break,
-                // regular character
-                c if is_qdtext(*c) => buffer.push(*c),
-                // Invalid character
-                c => {
-                    debug!(
-                        "Encountered an illegal character {} at location {}",
-                        *c as char,
-                        t.location().offset()
-                    );
-                    return None;
-                }
+            if challenge.scheme == XMatrix::SCHEME {
+                xmatrix = Some(challenge);
+                break;
             }
         }
-        Some(buffer)
-    })
-}
 
-fn parse_xmatrix_field<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<(String, Vec<u8>)> {
-    tokens.optional(|t| {
-        let name = parse_token(t).and_then(|name| {
-            let name = std::str::from_utf8(&name).ok()?.to_ascii_lowercase();
-            match name.as_str() {
-                "origin" | "destination" | "key" | "sig" => Some(name),
-                name => {
-                    debug!(
-                        "Encountered an invalid field name {} at location {}",
-                        name,
-                        t.location().offset()
-                    );
-                    None
-                }
-            }
-        })?;
+        let Some(xmatrix) = xmatrix else {
+            return Err(XMatrixParseError::NotFound);
+        };
 
-        if !t.token(&b'=') {
-            return None;
-        }
-
-        let value = parse_quoted(t).or_else(|| parse_token_with_colons(t))?;
-
-        Some((name, value))
-    })
-}
-
-fn parse_xmatrix<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<XMatrix> {
-    tokens.optional(|t| {
-        if !t.tokens(b"X-Matrix ") {
-            debug!("Failed to parse X-Matrix credentials, didn't start with 'X-Matrix '");
-            return None;
-        }
         let mut origin = None;
         let mut destination = None;
         let mut key = None;
         let mut sig = None;
 
-        for (name, value) in t.sep_by(|t| parse_xmatrix_field(t), |t| t.token(&b',')).as_iter() {
-            match name.as_str() {
+        for (name, value) in xmatrix.params {
+            match name {
```

Comment:
As per https://github.com/ruma/ruma/pull/1805/files#r1618914984:
```suggestion
            match name.to_lowercase() {
```

Reply:
It looks like the scheme should also be case-insensitive.

Reply:
Instead of allocating a string, we could use [`str::eq_ignore_ascii_case`](https://doc.rust-lang.org/std/primitive.str.html#method.eq_ignore_ascii_case).

Reply:
Done with my suggestion